Fix `at::native::view_as_real()` for ComplexHalf Tensors

malfet opened the pull request:
Add ComplexHalf case to toValueType, which fixes the logic how view_as_real and view_as_complex slices complex tensor to the floating point one, as it is used to generate tensor of random complex values, see:
https://github.com/pytorch/pytorch/blob/018b4d7abb894c3de21f9f10b0678c80a7b0a701/aten/src/ATen/native/DistributionTemplates.h#L200
Also add ability to convert python complex object to `c10::complex<at::Half>`

Add `torch.half` and `torch.complex32` to the list of `test_randn` dtypes

Fixes https://github.com/pytorch/pytorch/issues/43143

malfet committed:
Fix `at::native::view_as_real()` for ComplexHalf Tensors

Add ComplexHalf case to toValueType, which fixes the logic how view_as_real and view_as_complex slices the tensors for float, double and half types

Fixes https://github.com/pytorch/pytorch/issues/43143

anjali411 reviewed (changes requested):
thanks for the fix @malfet I think we should add a test [here](https://github.com/pytorch/pytorch/blob/018b4d7abb894c3de21f9f10b0678c80a7b0a701/test/test_torch.py#L19659) and [here](https://github.com/pytorch/pytorch/blob/018b4d7abb894c3de21f9f10b0678c80a7b0a701/test/test_torch.py#L11613) to verify this change. Since we have disabled `torch.chalf` dtype, we currently can't do that. 
```
>>> torch.tensor([1j], dtype=torch.complex32)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
RuntimeError: invalid type
```
One possibility is to disable `torch.chalf` for the next release but not disable it for the Nightly.

In addition, it might also be nice to update `torch.view_as_complex()` to support `torch.chalf` as well.

malfet committed:
Add ability to construct ComplexHalf tensors from Python complex values

Add `at::kComplexHalf' cases to `store_scalar` and `load_scalar` methods
Add `c10::complex<at::Half>::complex<at::Half>(c10::complex<double>)` constructor

malfet committed:
Add `torch.half` and `torch.complex32` to the list fo test_randn dtypes

malfet commented:
@anjali411 thanks for the feedback. I've added torch.half and torch.complex32 dtypes to test_randn, but the other change is a bit more involved, will publish it as a separate PR.

albanD reviewed (approved):
This one looks ok.
More changes are needed in other places as discussed offline but they are left for a future PR.

malfet committed:
Modernize code a bit
